---
layout:     post                    
title:      异常简单学              
subtitle:   异常的基本了解
date:       2018-02-05              
author:     Elijiah                    
header-img: img/timg (12).jpg   
catalog: true                       
tags:                               
    - 异常
---

## Hey

#什么是异常：

Java代码在运行时期发生的问题就是异常。

异常包含了错误的类型、原因以及位置

在Java中，把异常信息封装成了一个类。当出现了问题时，就会创建异常类对象并抛出异常相关的信息（如异常出现的位置、原因等）。

另外，在异常Exception类中，有一个子类要特殊说明一下，RuntimeException子类，RuntimeException及其它的子类只能在Java程序运行过程中出现。

我们再来观察Throwable类，能够发现与异常Exception平级的有一个Error，它是Throwable的子类，它用来表示java程序中可能会产生的严重错误。解决办法只有一个，修改代码避免Error错误的产生。

#异常的体系结构：

            Throwable（最顶层）

                    Error:出现的不能够处理的严重问题

                    Exception:可以处理的问题

#JVM默认处理方式

如果出现异常我们没有处理，jvm会帮我们进行处理，他会把异常的类型，原因还有位置显示在命令行并且还终止了程序，异常后面的代码将不在执行

捕获：Java中对异常有针对性的语句进行捕获，可以对出现的异常进行指定方式的处理

捕获异常格式

抛出去：当我们不想处理异常，或者没有能力处理的时候，我们可以抛出异常
使用关键字throws在方法声明处抛出

* public class ExceptionDemo2 {

	public static void main(String[] args) throws Exception {
	
		//method();
		function();
		
	}
	
	public static void function() throws Exception {
	
		FileWriter fw = new FileWriter("a.txt");
		
	}

	private static void method() {
	
		try {
		
			System.out.println(1);
			
			//System.out.println(2 / 0);
			
			System.out.println(2);
			
		} catch(ArithmeticException ae) {
		
			System.out.println("除数不能为0");
			
		}
		
		System.out.println(3);
	}

 }


try：该代码块中编写可能产生异常的代码。

catch：用来进行某种异常的捕获，实现对捕获到的异常进行处理。

使用一个try，多个catch：

1.多个catch可以有子父类关系

2.平级没有顺序关系

3.父类异常必须放到后面

finally的概述和应用场景

try {

	//需要被检测的语句。

}

catch(异常类 变量) { //参数。

	//异常的处理语句。

}

finally {

	//一定会被执行的语句。

}


 finally:组合try...catch使用，用于释放资源等收尾工作，无论try...catch语句如何执行，finally的代码一定会执行

#多异常处理

A:对代码进行异常检测，并对检测的异常传递给catch处理。对每种异常信息进行不同的捕获处理。

* void show(){ //不用throws 

	try{
		throw new Exception();//产生异常，直接捕获处理

	}catch(XxxException e){

 //处理方式	

	}catch(YyyException e){

 //处理方式	

	}catch(ZzzException e){

 //处理方式	

	}

 }

注意:这种异常处理方式，要求多个catch中的异常不能相同，并且若catch中的多个异常之间有子父类异常的关系，那么子类异常要求在上面的catch处理，父类异常在下面的catch处理。


#异常打印方式：

e. getMessage()：返回异常的原因

e.toString() ：返回异常的类型和原因

 printStackTrace():返回异常的类型，原因和位置




#1.1 编译时异常&运行时异常

A: 编译时期异常:是Exception的子类，非RuntimeExcpetion的子类，在编译时期必须处理

B:RuntimeException和他的所有子类异常,都属于运行时期异常。NullPointerException,ArrayIndexOutOfBoundsException等都属于运行时期异常，在编译时期可以自由选择处理或者不处理。

#运行时期异常的特点:

方法中抛出运行时期异常,方法定义中无需throws声明,调用者也无需处理此异常
运行时期异常一旦发生,需要程序人员修改源代码.


注意：如果抛出（throw）的是编译时期异常，必须在方法声明处抛出（throws）

throws:处理异常的一种方式，把异常抛出，由调用者来处理

throw:制造异常的方式,并且结束方法

#以后开发处理异常的原则：

最底层抛（throws），最顶层必须处理（try...catch），尽可能的释放资源

#如何自定义一个异常类：

非常简单，写一个类去继承Exception或者RuntimeException，然后实现多个构造即可

* public class MyException extends /*RuntimeException*/ Exception{

	public MyException() {

		super();

	}

	public MyException(String message) {

		super(message);

	}
	/*public MyException() {

		super();

	}
	
	public MyException(String s) {

		super(s);

	}*/
 }

作者：Elijiah

來源：自己总结的学习笔记

著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。
